### PR TITLE
[FLINK-3754][Table]Add a validation phase before construct RelNode using TableAPI

### PR DESCRIPTION
**Pull request by yjshen**
This PR aims at adding an extra phase of **validation** for plans generated from Table API, matches the functionality of Calcite's Validator that are called during we execute an query expressed in SQL String.

In order to do this, I inserted a new layer between TableAPI and `RelNode` construction: The `Logical Plan`.

And the main procedure of validation and RelNode construction work as follows:

1. Constructing logical plan in flight
2. When we are about to execute the plan:
3. Do resolution using table/dataset/datastream schema
4. Do validation on the type annotated logical plan tree
5. Translate logical plan into `RelNode` once we successfully finished validation.

### CODE REVIEW

**Comment by fhueske** [comment]
Thanks @yjshen for working on this issue! Unified validation and exceptions are a big improvement, IMO. I'll try to have a look soon.

@twalthr, can you have a look at the changes done on the expressions?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 1]
```diff
@@ -199,12 +151,12 @@ class Table(
...
-
-    relBuilder.push(relNode)
-    relBuilder.filter(predicate.toRexNode(relBuilder))
-    
-    new Table(relBuilder.build(), tableEnv)
+  def filter(predicate: Expression): Table = withPlan {
+//    logicalPlan match {
+//      case j: Join => j.copy(condition = Some(predicate))
+//      case o => Filter(predicate, logicalPlan)
+//    }
```
Comment (yjshen) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala]: Will remove this the next commit.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+case class Union(left: LogicalNode, right: LogicalNode) extends BinaryNode {
+  override def output: Seq[Attribute] = left.output
+
+  override def toRelNode(relBuilder: RelBuilder): RelBuilder = {
+    left.toRelNode(relBuilder)
+    right.toRelNode(relBuilder)
+    relBuilder.union(true)
+  }
+}
```
Comment (yjshen) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: Need to override the `resolved` field to check equality of `left` and `right` output schema. Will do in next commit.

**Comment by yjshen** [comment]
For ease of review, I would like to explain this RP with more details.

While table api are called to construct a query, it was first constructed as a operator tree, `LogicalNode` is used to express the tree node. Therefore, after we finished constructing a query and about to translate it into a dataset/datastream program, the constructed logical plan looks like:
```
Aggregate(List of group by, List of aggregate)
+- Filter (condition expression)
    +- Union
        +- Project (select column expression) 
            +- TableScan(DS1) 
        +- Project (select column expression)
            +- TableScan(DS2)
```
At this time, only the leaf node: TableScan (`CatalogNode` in current implementation) is equipped with type information, in order to do plan validation, we need to **annotate** the full logical plan with type information first and use type information to do **validate** works.

**Comment by yjshen** [comment]
The type annotation work is done from bottom to top:

Firstly, we know each schema of the two input, and we know `List[] expression` in `Project` are used to manipulate one row of table data as input and output one value per expression, therefore, we can infer the the output schema of `Project` (in the current impl this was expressed as: `def output: Seq[Attribute]`) if we know each expressions `dataType`. 

For example, `Add`'s dataType is same as it's input, `Or`'s dataType is always `Boolean`, `pow(a, b)`'s dataType is always `Double`, however, if and only if we understand all kinds of expressions, we are able to infer its `dataType`. The main problems here is we only have `Call`(Unresolved Function) generated during expression construction, therefore, we should resolve them first into solid `Expression`s. `FunctionCatalog` is introduced here for a mapping from `FunctionName -> Expression`, we can easily finish the translation work as we look up `catalog`.

**Comment by fhueske** [comment]
Thanks for the additional information @yjshen! I'm a bit behind with PR reviews, but will definitely have a look begin of next week. Thanks, Fabian

**Comment by yjshen** [comment]
@fhueske, thanks for your job. Looking forward to more discussions on this :)

**Comment by twalthr** [comment]
Thanks for the contribution @yjshen. I will also have a look at it tomorrow.

**Comment by twalthr** [comment]
@yjshen I looked through the code changes. I was quite impressed that your PR touches nearly every class of the current API.

Basically your changes seem to work, however, I'm not sure if we want to implement the validation phase for every scalar function ourselves. Actually, Calcite already comes with type inference, type checking and validation capabilities. I don't know if we want to reinvent the wheel at this point. Your approach inserts a layer under the Table API for doing the validation. However, instead, this layer could also translate the plan into a SQL tree (on top of RelNodes). We could then let Calcite do the work of validation.

This could also solve another problem that I faced when working on FLINK-3580. If you take a look at `StandardConvertletTable` of Calcite, you see that Calcite also does some conversions which we also need to implement ourselves if we do not base the Table API on top of SQL.

We need to discuss how we want to proceed. Both solution are not perfect.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala [hunk 3]
```diff
@@ -355,4 +380,26 @@ object TableEnvironment {
...
+    */
+  class PlanPreparation(val env: TableEnvironment, val logical: LogicalNode) {
+
+    lazy val resolvedPlan: LogicalNode = env.getValidator.resolve(logical)
+
+    def validate(): Unit = env.getValidator.validate(resolvedPlan)
+
+    lazy val relNode: RelNode = {
+      env match {
+        case _: BatchTableEnvironment =>
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala]: Why do you distinguish here? It's the same code, no?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/expressions/Expression.scala [hunk 4]
```diff
@@ -17,13 +17,34 @@
...
-abstract class Expression extends TreeNode[Expression] { self: Product =>
-  def name: String = Expression.freshName("expression")
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.api.table.trees.TreeNode
+import org.apache.flink.api.table.validate.ExprValidationResult
+
+abstract class Expression extends TreeNode[Expression] {
+  /**
+    * Returns the [[TypeInformation]] for evaluating this expression.
+    * It is sometimes available until the expression is valid.
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/expressions/Expression.scala]: +not ?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/expressions/Expression.scala [hunk 5]
```diff
@@ -17,13 +17,34 @@
...
+    */
+  lazy val valid: Boolean = childrenValid && validateInput().isSuccess
+
+  def childrenValid: Boolean = children.forall(_.valid)
+
+  /**
+    * Check input data types, inputs number or other properties specified by this expression.
+    * Return `ValidationSuccess` if it pass the check,
+    * or `ValidationFailure` with supplement message explaining the error.
+    * Note: we should only call this method until `childrenValidated == true`
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/expressions/Expression.scala]: `childrenValidated` -> `childrenValid`?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/table/test/ScalarFunctionsTest.scala [hunk 6]
```diff
@@ -215,7 +215,7 @@ class ScalarFunctionsTest {
 
   }
 
-  @Test
+  @Ignore
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/table/test/ScalarFunctionsTest.scala]: Why did you exclude these tests?

**Comment by fhueske** [comment]
Hi @yjshen, thanks for your patience. I also finished a first pass over the PR. 

I'd like to propose a third alternative, in addition to the custom validation phase (this PR) and generating `SqlNode`s and using Calcite's validator. Both approaches would mean that the validation happens before the logical plan is translated into a `RelNode`. I think it would be good to eagerly check each method call of the Table API. This would make debugging easier, because exceptions would be thrown where the error is caused. Please correct me if I am wrong, but I think we would not lose validation coverage compared to the coverage this PR if we do eager validation? It might also be easier, because we do not need the recursive operator traversal (still the expression traversal though). Maybe we can even directly translate to `RelNode`s after validation, just like we do right now. I think a lot of this PR could be used for eager validation, not sure if it would be easily possible with the `SqlNode` validation approach. 
What do you think about eagerly validation, @yjshen and @twalthr?

While reviewing the PR, I noticed that some classes seem to be partially derived from Spark's code base (e.g., `TreeNode` and `RuleExecutor`). I know there are some common patterns that apply in optimizers, but it is good style to give credit to the original source code. 
Can you list which classes are derived from Spark code and add a comment to them pointing to the source of the code?

Thanks, Fabian

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/expressions/Expression.scala (hunk 5 above)
Comment (yjshen) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/expressions/Expression.scala]: Ah, yes!

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/table/test/ScalarFunctionsTest.scala (hunk 6 above)
Comment (yjshen) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/table/test/ScalarFunctionsTest.scala]: `exp` `log` `pow` and `ln` should have `Double` as input. What I was thinking is, we should add some extra `type coercion` rules and add an `cast` when we can do it safely(when an expression is asking a `Double` but we provide a `Int`), for example, `Byte` to `Double`, `Long` to `Double` and so on, and rewrite the expression as `pow(cast(x, Double))`

**Comment by yjshen** [comment]
> I think it would be good to eagerly check each method call of the Table API. This would make debugging easier, because exceptions would be thrown where the error is caused. Please correct me if I am wrong, but I think we would not lose validation coverage compared to the coverage this PR if we do eager validation? It might also be easier, because we do not need the recursive operator traversal (still the expression traversal though). Maybe we can even directly translate to RelNodes after validation, just like we do right now. I think a lot of this PR could be used for eager validation, not sure if it would be easily possible with the SqlNode validation approach. 

Regarding the eager validation you mentioned, I think that could be accomplished by calling `validate()` each time I am constructing another `Table`, in other words, each time I am calling a `Table api`, changing the current code from:
``` scala
class Table(
    private[flink] val tableEnv: TableEnvironment,
    private[flink] val planPreparation: PlanPreparation) {

  def this(tableEnv: TableEnvironment, logicalPlan: LogicalNode) = {
    this(tableEnv, new PlanPreparation(tableEnv, logicalPlan))
  }
```
to 
``` scala
def this(tableEnv: TableEnvironment, logicalPlan: LogicalNode) = {
    this(tableEnv, {
      val pp = new PlanPreparation(tableEnv, logicalPlan)
      pp.validate()
      pp
    })
```
and also add an additional flag annotate the logical node as `validate` and therefore avoid the `recursive logical plan traversal` would be enough.  Do I understand your idea correctly?
On the other hand, I prefer to postponed `RelNode` construction until we are going to run the query on Flink, since we only need `RelNode` at that time.

> While reviewing the PR, I noticed that some classes seem to be partially derived from Spark's code base (e.g., TreeNode and RuleExecutor). I know there are some common patterns that apply in optimizers, but it is good style to give credit to the original source code. 
Can you list which classes are derived from Spark code and add a comment to them pointing to the source of the code?

OK, I will do a more detailed scan and give credit to all the original source code.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala (hunk 3 above)
Comment (yjshen) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala]: Yes, it shouldn't be distinguished. The difference should exist inside validator's rule set.

**Comment by yjshen** [comment]
@twalthr, translate table API call to `SqlNode` and use `validator` in Calcite seems another reasonable solution, maybe it's hard to do `eager validation` as @fhueske proposed? 
Yes, I agree we should discuss which way is preferred solution :)

**Comment by yjshen** [comment]
Hi @fhueske @twalthr, thanks for the detailed review work! 👍 
I've left some comment above to express my opinion, looking forward to more discussions on this :)

**Comment by fhueske** [comment]
If we decide for eager validation, we can also simplify the code a bit, right? 

I might be wrong but we could remove `PlanPreparation` and the recursive validation of `LogicalNode` since the inputs of an operator would have been already validated. The checks of `Validator.validate()` could be moved to the respective `LogicalNode`. Not sure if we would still need the `RuleExecutor`.

**Comment by yjshen** [comment]
@fhueske If that's the preferred way, I will try to do some simplifications and see.

**Comment by yjshen** [comment]
Will start to work on eager validation now. :)

**Comment by fhueske** [comment]
Great, thanks!

**Pull request closed without merging**